### ISSUE TITLE
On AMD GPUs (ROCm 5.7-6.2), cannot backpropagate loss tensor containing more than `2e8` elements

### ISSUE THREAD

**Issue by amorehead** [pytorch/pytorch#136291]
### 🐛 Describe the bug

In the following code, I am encountering a `HIP error` during backpropagation that is thrown when executing `bond_losses[atompair_mask].mean()`. I initially suspected (based on prior experience with e.g., AMD GPUs, namely the MI250x line) that this was caused by hardware/driver limitations for ROCm (n.b., issue reproduced with versions 5.7, 6.1, and 6.2), and adding the `if` clause below confirmed my suspicions since the issue during backpropagation has now disappeared as a result. I've made a pull request into [another GitHub repository](https://github.com/lucidrains/alphafold3-pytorch/pull/271) that makes use of this workaround for AMD GPUs.

```python
...

MAX_ELEMENTS_FOR_BACKPROP = int(2e8)

bond_losses = F.mse_loss(denoised_cdist, normalized_cdist, reduction = 'none')
bond_losses = bond_losses * loss_weights

if atompair_mask.sum() > MAX_ELEMENTS_FOR_BACKPROP:
      # randomly subset the atom pairs to supervise to avoid HIP error during backprop
      
      flat_atompair_mask_indices = torch.arange(atompair_mask.numel(), device=self.device)[atompair_mask.view(-1)]
      num_true_atompairs = flat_atompair_mask_indices.size(0)
      
      num_atompairs_to_ignore = num_true_atompairs - MAX_ELEMENTS_FOR_BACKPROP
      ignored_atompair_indices = flat_atompair_mask_indices[torch.randperm(num_true_atompairs)[:num_atompairs_to_ignore]]
      
      atompair_mask.view(-1)[ignored_atompair_indices] = False

bond_loss = bond_losses[atompair_mask].mean()  # <- without the if-clause above, this is the error-triggering line for large inputs
```

It's also worth noting that `2e8` isn't a precise threshold where this issue occurs. In practice, an `atompair_mask` of shape `[B, N, N]`, where `B=4` and `N≈8,600`, is in the range of tensor sizes where I started seeing this issue arise. To be on the safe side, I decided on setting `MAX_ELEMENTS_FOR_BACKPROP=int(2e8)` since this is close to setting the maximum size at `[4, 7000, 7000]`.

### Versions

Collecting environment information...
PyTorch version: 2.3.0a0+git96dd291
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 6.2.41133-dd7f95766

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 18.0.0git (https://github.com/RadeonOpenCompute/llvm-project roc-6.2.0 24292 26466ce804ac523b398608f17388eb6d605a3f09)
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.14.21-150500.55.49_13.0.57-cray_shasta_c-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: AMD Instinct MI250X (gfx90a:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: 6.2.41133
MIOpen runtime version: 3.2.0
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               128
On-line CPU(s) list:                  0-127
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 7A53 64-Core Processor
CPU family:                           25
Model:                                48
Thread(s) per core:                   2
Core(s) per socket:                   64
Socket(s):                            1
Stepping:                             1
Frequency boost:                      enabled
CPU max MHz:                          3541.0149
CPU min MHz:                          1500.0000
BogoMIPS:                             3992.57
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                       AMD-V
L1d cache:                            2 MiB (64 instances)
L1i cache:                            2 MiB (64 instances)
L2 cache:                             32 MiB (64 instances)
L3 cache:                             256 MiB (8 instances)
NUMA node(s):                         4
NUMA node0 CPU(s):                    0-15,64-79
NUMA node1 CPU(s):                    16-31,80-95
NUMA node2 CPU(s):                    32-47,96-111
NUMA node3 CPU(s):                    48-63,112-127
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] adam-atan2-pytorch==0.0.12
[pip3] alphafold3-pytorch-lightning-hydra==0.5.25
[pip3] frame-averaging-pytorch==0.1.2
[pip3] lion-pytorch==0.2.2
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] optree==0.9.1
[pip3] pytorch-lightning==2.4.0
[pip3] rotary-embedding-torch==0.8.3
[pip3] torch==2.3.0a0+git96dd291
[pip3] torch-geometric==2.6.0
[pip3] torchmetrics==1.4.2
[pip3] torchtyping==0.1.5
[pip3] torchvision==0.18.0a0+68ba7ec
[pip3] triton==2.3.0
[conda] Could not collect

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

**Comment by amorehead** [comment]
An update: I've noticed a similar `HIP error` (invalid configuration) getting triggered when I feed an input tensor with more than `int(1e8)` elements into `torch.nn.LayerNorm()`.

**Comment by hongxiayang** [comment]
> An update: I've noticed a similar `HIP error` (invalid configuration argument) getting triggered when I feed an input tensor with more than `int(1e8)` elements into `torch.nn.LayerNorm()`.

@amorehead Can you post a simple reproducible script so that can reproduce and debugging the issue? Thanks.

**Comment by amorehead** [comment]
@hongxiayang, sure thing. Below shows my current workaround for this issue.

```python
# in my nn.Module's init function...
from einops.layers.torch import Rearrange

LinearNoBias = partial(nn.Linear, bias=False)
to_attn_bias_linear = LinearNoBias(dim_pairwise, heads)
nn.init.zeros_(to_attn_bias_linear.weight)

self.to_attn_bias_norm = nn.LayerNorm(dim_pairwise)
self.to_attn_bias = nn.Sequential(to_attn_bias_linear, Rearrange("b ... h -> b h ..."))

# in my nn.Module's forward function...
MAX_CONCURRENT_TENSOR_ELEMENTS = int(1e8)  # my workaround
dtype, device = pairwise_repr.dtype, pairwise_repr.device

if pairwise_repr.numel() > MAX_CONCURRENT_TENSOR_ELEMENTS:
# create a stub tensor and normalize it to maintain gradients to `to_attn_bias_norm`
stub_pairwise_repr = torch.zeros((b, dp), dtype=dtype, device=device)
stub_attn_bias_norm = self.to_attn_bias_norm(stub_pairwise_repr) * 0.0

# adjust `attn_bias_norm` dimensions to match `pairwise_repr`
attn_bias_norm = pairwise_repr + stub_attn_bias_norm[:, None, None, :]

# apply bias transformation
attn_bias = self.to_attn_bias(attn_bias_norm) + attn_bias
else:
attn_bias = self.to_attn_bias(self.to_attn_bias_norm(pairwise_repr)) + attn_bias
```
where `pairwise_repr` is a `Float["b n n dp"]` tensor and `attn_bias` is a `Float["b n n"]` tensor.

**Comment by hongxiayang** [comment]
@amorehead  Can you simplify your script so that it can run independently (i.e. without any other dependencies except `torch` )  and it will throw the HIP error (invalid configuration argument)?

**Comment by amorehead** [comment]
@hongxiayang, apologies. Here's a simple script with which to reproduce the error with ROCm 6.2 and PyTorch 2.3.0. Note that I'm running this bug testing script (as well as my original training script) within a Singularity container on a SLURM cluster. The Singularity container is running a pre-built ROCm-official [Docker image](https://hub.docker.com/layers/rocm/pytorch/rocm6.2_ubuntu22.04_py3.10_pytorch_release_2.3.0/images/sha256-931d3e3dcebe6c6fab84adf16cfca3e1d1449100df7c881a46fccd06f6c9bc1c?context=explore) that was converted into a compatible format for Singularity using the command `singularity pull $TARGET_DOCKER_IMAGE`.

```python
from functools import partial

import torch
import torch.nn as nn

MAX_CONCURRENT_TENSOR_ELEMENTS = int(1e20)  # NOTE: set to 1e8 for my workaround


class ReproducePyTorchIssue136291(nn.Module):
    """Reproduce PyTorch issue #136291."""

    def __init__(
        self,
        b=16,
        n=1000,
        dp=16,
        heads=16,
        dtype=torch.float32,
        device=torch.device("cuda"),
    ):
        super(ReproducePyTorchIssue136291, self).__init__()
        self.b = b
        self.n = n
        self.dp = dp
        self.heads = heads
        self.dtype = dtype
        self.device = device

        LinearNoBias = partial(nn.Linear, bias=False)
        self.to_attn_bias_linear = LinearNoBias(dp, heads).to(device=device)
        nn.init.zeros_(self.to_attn_bias_linear.weight)

        self.to_attn_bias_norm = nn.LayerNorm(dp).to(device=device)
        self.to_attn_bias = self.to_attn_bias_linear

    def forward(self, pairwise_repr):
        pairwise_repr = pairwise_repr.to(dtype=self.dtype, device=self.device)

        if pairwise_repr.numel() > MAX_CONCURRENT_TENSOR_ELEMENTS:
            stub_pairwise_repr = torch.zeros(
                (self.b, self.dp), dtype=self.dtype, device=self.device
            )
            stub_attn_bias_norm = self.to_attn_bias_norm(stub_pairwise_repr) * 0.0

            attn_bias_norm = pairwise_repr + stub_attn_bias_norm[:, None, None, :]

            attn_bias = self.to_attn_bias(attn_bias_norm).view(self.b, self.heads, self.n, self.n)
        else:
            attn_bias = self.to_attn_bias(self.to_attn_bias_norm(pairwise_repr)).view(
                self.b, self.heads, self.n, self.n
            )

        return attn_bias


if __name__ == "__main__":
    model = ReproducePyTorchIssue136291()
    pairwise_repr = torch.randn(16, 1000, 1000, 16, device="cuda")
    output = model(pairwise_repr)

    print(output.shape)
```

**Comment by amorehead** [comment]
Hi, @hongxiayang. Have you by chance found anything that might be causing this issue above?

**Comment by hongxiayang** [comment]
@naromero77amd  @amorehead  It seems this might be similar to the issue I fixed a while ago. Please check this PR: https://github.com/pytorch/pytorch/pull/130994 for ideas of how to fix it.

**Comment by naromero77amd** [comment]
@amorehead Thanks for the small reproducer. I was able to reproduce the issue on an MI200.

I do see a hipLaunchKernel with a large argument:
```
:3:hip_module.cpp           :677 : 25417836116558 us: [pid:2778261 tid:0x7f770f1a6740] [32m hipLaunchKernel ( 0x7f76fa8f1550, {144000000,1,1}, {64,4,1}, 0x7ffc0c98a9a0, 24, stream:<null> ) [0m
:3:devprogram.cpp           :2648: 25417836116866 us: [pid:2778261 tid:0x7f770f1a6740] Using Code Object V5.
```
Will triage more next week.

**Comment by amorehead** [comment]
@naromero77amd, thanks for letting me know!

**Comment by hongxiayang** [comment]
@amorehead  I have a [PR](https://github.com/pytorch/pytorch/pull/144007) now. If it is possible, please verify and let me know.

**Comment by amorehead** [comment]
@hongxiayang, thanks a ton for creating this PR! Is there a streamlined way I can pull these changes down in the form of a pre-built PyTorch Docker image? For reference, the base Docker image I am currently using in my training environment is `rocm/pytorch:rocm6.2_ubuntu22.04_py3.10_pytorch_release_2.3.0`.

**Comment by jithunnair-amd** [comment]
> @hongxiayang, thanks a ton for creating this PR! Is there a streamlined way I can pull these changes down in the form of a pre-built PyTorch Docker image? For reference, the base Docker image I am currently using in my training environment is `rocm/pytorch:rocm6.2_ubuntu22.04_py3.10_pytorch_release_2.3.0`.

@amorehead Once the PRs with the fixes are merged, they should be part of the next day's [rocm/pytorch-nightly:latest](https://hub.docker.com/r/rocm/pytorch-nightly/tags) docker image.

**Comment by amorehead** [comment]
Thanks, @jithunnair-amd!

**Comment by hongxiayang** [comment]
@amorehead The https://github.com/pytorch/pytorch/pull/144007 is merged, and the issue is automatically closed by the bot. Please check with the [rocm/pytorch-nightly:latest](https://hub.docker.com/r/rocm/pytorch-nightly/tags) and let us know.

**Comment by hongxiayang** [comment]
> [@hongxiayang](https://github.com/hongxiayang), thanks again! I've tested this latest ROCm nightly PyTorch 2.5.1 image (with PyTorch Lightning/Fabric 2.5.1 and Triton 3.1.0 as well), and everything seems to be working smoothly on my ROCm SLURM cluster.

Happy to hear this from you, @amorehead , and thank you for confirming that it works for you in slurm environment now.
